### PR TITLE
Return NotImplemented from tensor arithmetic operators

### LINKED ISSUE #26333: torch.Tensor reverse operators (__rmul__) should return NotImplemented for unsupported types
Currently if the the argument of `__rmatmul__` is not a Tensor, PyTorch throws `*** TypeError: matmul(): argument 'other' (position 1) must be Tensor, not Vec`

Python data model asks that it return `NotImplemented` in such case. This allows interpreter to dispatch op to the proper implementation.
(ie see https://docs.python.org/3/library/numbers.html#implementing-the-arithmetic-operations)

Current behavior makes it hard to implement custom types. IE
```
class FactoredMatrix:
  def __init__(self, mat):
    self.mat = mat
  def __matmul__(self, other):
    return 0
  def __rmatmul__(self, other):
    return 1
x = torch.ones((2,2))
print(FactoredMatrix(x) @ x)  # works
print(x @ FactoredMatrix(x))  # fails
```

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #26333

This effectively rewrites all infix arithmetic operators *op* as:
```python
def __op__(self, other):
    try:
        return self.op(other)
    except TypeError:
        return NotImplemented
```

Where `TypeError` is raised from the argument parser when `other` is not a `Tensor`. This should be okay, so long as `TypeError` isn't raised by the function for any other reasons. I couldn't find any examples where this was an issue.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Nifty!

**Comment by ezyang** [comment]
Waiting on CI

**Comment by ezyang** [comment]
Gosh, this PR conflicts a lot for some reason.

**Comment by ezyang** [comment]
Oh, you didn't grant maintainers rights to update the PR, so the bot can't update it. Could you please rebase again to clear up the errors?

**Comment by peterbell10** [comment]
>Oh, you didn't grant maintainers rights to update the PR, so the bot can't update it.

That's odd, "Allow edits from maintainers." is shown a selected on my screen. Could it be an issue with submitting from the Quansight repo?

cc @rgommers

**Comment by rgommers** [comment]
> That's odd, "Allow edits from maintainers." is shown a selected on my screen. Could it be an issue with submitting from the Quansight repo?

Weirdly enough, I think it is. Just tried on PR 25629 and that failed as well.

Our repository settings are a bit unusual, I'll look into it.

Tried adding `pytorchbot` as a collaborator to fix it, but that requires it to accept an invitation:

image

I'll try to look for a more structural fix, because if it is repository settings then it will happen on other repos as well.

**Comment by ezyang** [comment]
I had pytorchbot accept the invitation.

**Comment by ezyang** [comment]
Didn't work :/

**Comment by rgommers** [comment]
Hmm, then I have no idea. Will have to do some trial and error on a dummy repo. @peterbell10 for next PRs I suggest to open them from your own fork until we figure this out.

**Comment by gchanan** [comment]
should we be returning NotImplemented for the inplace functions?  That doesn't seem correct.

Also, this doesn't work fork for functions that are written in native_functions.yaml and aren't explicitly bound.

**Comment by ezyang** [comment]
The docs specify that it is valid for inplace functions too:

> Special value which should be returned by the binary special methods (e.g. __eq__(), __lt__(), __add__(), __rsub__(), etc.) to indicate that the operation is not implemented with respect to the other type; may be returned by the in-place binary special methods (e.g. __imul__(), __iand__(), etc.) for the same purpose. Its truth value is true.

Rereviewing this diff, I regret not asking for tests though...

**Pull request closed without merging**
